central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -59,3 +123,37 @@ func (ds *datastoreImpl) GetBatch(ctx context.Context, ids []string) ([]*storage
 
 	return collections, nil
 }
+
+func (ds *datastoreImpl) AddCollection(ctx context.Context, collection *storage.ResourceCollection) error {
+
+	// add to graph first to detect any cycles
+	err := ds.addCollectionToGraph(ctx, collection)
+	if err != nil {
+		return err
+	}
+
+	err = ds.storage.Upsert(ctx, collection)
+	if err != nil {
+		// if we fail to upsert, update the graph
+		deleteErr := ds.deleteCollectionFromGraph(ctx, collection.GetId())
+		if deleteErr != nil {
+			log.Errorf("Failed to delete collection, might result in bad state (%v)", deleteErr)
```

Comment:
This error isn't about deleting collection, it's about the local state being inconsistent. The collection itself doesn't exist in the DB so a restart would fix it.

Depending on what the id is (guid vs auto-increment id), then it might also not be an issue. Because if the id is a guid the chances are low that the next upsert will have the exact same collection id.

Reply:
This is kind of the same issue as above, theoretically this issue should never occur since we know we just added the id to the graph, so removing it from the graph subsequently should never fail given the constraints of the function